src/util/coordinate_calculation.cpp

```diff
@@ -381,6 +381,218 @@ Coordinate difference(const Coordinate lhs, const Coordinate rhs)
+//             IEEE Trans. PAMI, Vol. 13, pages 1115-1138, (1991)
+
+// The method is based on the minimization of the function
+
+//              sum [(x-a)^2 + (y-b)^2 - R^2]^2
+//          F = -------------------------------
+//                  sum [(x-a)^2 + (y-b)^2]
+
+// This method is more balanced than the simple Kasa fit.
+
+// It works well whether data points are sampled along an entire circle or
+// along a small arc.
+
+// It still has a small bias and its statistical accuracy is slightly
+// lower than that of the geometric fit (minimizing geometric distances),
+// but slightly higher than that of the very similar Pratt fit.
+// Besides, the Taubin fit is slightly simpler than the Pratt fit
+
+// It provides a very good initial guess for a subsequent geometric fit.
+
+//   Nikolai Chernov  (September 2012)
+//   http://people.cas.uab.edu/~mosya/cl/CircleFitByTaubin.cpp
+bool circleCenterTaubin(const std::vector<Coordinate> &coords,
+                        std::pair<double, double> &center,
+                        double &estimated_radius)
+{
+    // Compute mean and covariances with the two-pass algorithm
+    double meanX = 0., meanY = 0.;
+    for (const auto &c : coords)
+    {
+        // Use the equirectangular projection
+        const auto x = static_cast<double>(toFloating(c.lon));
+        const auto y = static_cast<double>(toFloating(c.lat));
+        meanX += x;
+        meanY += y;
+    }
+    meanX /= coords.size();
+    meanY /= coords.size();
+
+    // Compute moments for centered x, y and z = x^2+y^2 coordinates
+    double Mxx = 0., Myy = 0., Mxy = 0., Mxz = 0., Myz = 0., Mzz = 0.;
+    for (auto &c : coords)
+    {
+        const auto Xi = static_cast<double>(toFloating(c.lon)) - meanX;
+        const auto Yi = static_cast<double>(toFloating(c.lat)) - meanY;
+        const auto Zi = Xi * Xi + Yi * Yi;
+
+        Mxx += Xi * Xi;
+        Myy += Yi * Yi;
+        Mzz += Zi * Zi;
+        Mxy += Xi * Yi;
+        Mxz += Xi * Zi;
+        Myz += Yi * Zi;
+    }
+
+    Mxx /= coords.size();
+    Myy /= coords.size();
+    Mxy /= coords.size();
+    Mxz /= coords.size();
+    Myz /= coords.size();
+    Mzz /= coords.size();
+
+    // Compute coefficients of the characteristic polynomial
+    const auto Mz = Mxx + Myy;
+    const auto cov_xy = Mxx * Myy - Mxy * Mxy;
+    const auto var_z = Mzz - Mz * Mz;
+    const auto c3 = 4. * Mz;
+    const auto c2 = -3. * Mz * Mz - Mzz;
+    const auto c1 = var_z * Mz + 4. * cov_xy * Mz - Mxz * Mxz - Myz * Myz;
+    const auto c0 = Mxz * (Mxz * Myy - Myz * Mxy) + Myz * (Myz * Mxx - Mxz * Mxy) - var_z * cov_xy;
+
+    // Find the root of the characteristic polynomial using Newton's method starting at x=0.
+    // It is guaranteed to converge to the smallest characteristic value
+    double x = 0., y = c0;
+    const auto eps = sqrt(std::numeric_limits<double>::epsilon());
+    for (int iter = 0; iter < 5; ++iter)
+    {
+        const auto dy = c1 + x * (2. * c2 + 3. * c3 * x);
```

Comment:
didn't check do to the quite complex math. but can `dy` ever be equal to `0`?

Reply:
yes, it is possible if x corresponds to a polynom extremum point, this is allowed case of a non-circular shape and handled by `std::isfinite` below